### PR TITLE
ci: skip riot runs based on diff from last successful run

### PR DESCRIPTION
**Pull request by emmettbutler**
This pull request reduces the amount of duplicated work done in CI on pull request builds by skipping a given riot job based on the changes since the last time that job succeeded.

The approach taken here reduces the amount of time spent waiting for jobs to pass after updating a branch from trunk, ensuring that only those jobs that might have been affected by the update are run.

This logic is supported by CircleCI's `save_cache` and `restore_cache` directives. The cache key includes the branch name, job name, and node number to ensure uniqueness for each branch/job/node combination. `restore_cache` uses a prefix match on this key format.

No new skips are applied on `*.x` branches, which mitigates much of the risk of this change.

Here's a detailed description of the logical flow.

1. A commit with SHA `A` is pushed to a new branch, changing `contrib/cherrypy/middleware.py`.
2. As on current 1.x, CI jobs are selected based on the commit's diff. The `cherrypy` job starts.
3. The `restore_cache` step in the `cherrypy` job finds no cache.
4. The `cherrypy` job succeeds and writes a file to the CI runner's filesystem containing the Git hash of the current HEAD, `A`. This is the "latest successful commit" for the `cherrypy` job on this branch.
5. The `save_cache` step caches this file under a key that includes the branch name, `cherrypy`, and the epoch timestamp.
6. Another commit with SHA `B` is pushed to the branch, changing `contrib/falcon/patch.py`. The branch now contains changes to `falcon` and `cherrypy`.
7. The dynamic configuration generator selects the `falcon` and `cherrypy` jobs to run.
8. In the `falcon` job, the `restore_cache` step finds no cache at the key `*-falcon-*`. In the `cherrypy` job, this step finds a cache.
9. The `falcon` job runs similarly to steps 4 and 5. The `cherrypy` job opens the file it found in the cache and passes the resulting Git SHA to `needs_testrun.py`. This script runs `git diff B A`, finds that none of the changes in that diff affect `cherrypy`, and exits before running `riot`.

**Why not do this check during the setup step?**
The new diff check added in this pull request requires access to the CircleCI cache on a per-job basis, which is only available from inside each job.

**How was this tested?**
Commit https://github.com/DataDog/dd-trace-py/pull/6600/commits/c6eaed7b47f45c4b358f67d37935478dd3e06fca made a change to `dbapi`. I observed its CI triggering all of the related jobs, and running `riot` on each. The `psycopg` job had an ephemeral failure, and the rest succeeded. Commit https://github.com/DataDog/dd-trace-py/pull/6600/commits/7e469f39de5289e7efd727f97fd504efcd16cc1a did not change any tested code. Its CI ran all of the `dbapi` jobs, and all of them except `psycopg` skipped their `riot` runs and exited early. The `psycopg` job ran its `riot` command.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: scripts/needs_testrun.py [hunk 1]
```diff
@@ -58,43 +58,36 @@ def get_merge_base(pr_number: int) -> str:
 
 
 @cache
-def get_changed_files(pr_number: int) -> t.Set[str]:
+def get_changed_files(pr_number: int, sha: str = "") -> t.Set[str]:
```
Comment (P403n1x87) [inline scripts/needs_testrun.py:61]:
```suggestion
def get_changed_files(pr_number: int, sha: t.Optional[str] = None) -> t.Set[str]:
```

File: scripts/needs_testrun.py [hunk 2]
```diff
@@ -58,43 +58,36 @@ def get_merge_base(pr_number: int) -> str:
...
-                    "HEAD",
-                    get_merge_base(pr_number),
-                ]
-            )
-            .decode("utf-8")
-            .strip()
-            .splitlines()
-        )
+    rest_check_failed = False
+    if not sha:
```
Comment (P403n1x87) [inline scripts/needs_testrun.py:74]:
```suggestion
    if sha is None:
```

File: scripts/needs_testrun.py [hunk 3]
```diff
@@ -58,43 +58,36 @@ def get_merge_base(pr_number: int) -> str:
...
+        try:
+            url = f"https://api.github.com/repos/datadog/dd-trace-py/pulls/{pr_number}/files"
+            headers = {"Accept": "application/vnd.github+json"}
+            return {_["filename"] for _ in json.load(urlopen(Request(url, headers=headers)))}
+        except Exception:
+            rest_check_failed = True
+            LOGGER.warning("Failed to get changed files from GitHub API")
+
+    if sha or rest_check_failed:
```
Comment (P403n1x87) [inline scripts/needs_testrun.py:83]:
```suggestion
    if sha is not None or rest_check_failed:
```

File: scripts/needs_testrun.py [hunk 4]
```diff
@@ -58,43 +58,36 @@ def get_merge_base(pr_number: int) -> str:
...
+        diff_base = sha or get_merge_base(pr_number)
+        LOGGER.info("Checking changed files against commit %s", diff_base)
+        return set(check_output(["git", "diff", "--name-only", "HEAD", diff_base]).decode("utf-8").strip().splitlines())
 
 
 @cache
-def needs_testrun(suite: str, pr_number: int) -> bool:
+def needs_testrun(suite: str, pr_number: int, sha: str = "") -> bool:
```
Comment (P403n1x87) [inline scripts/needs_testrun.py:90]:
```suggestion
def needs_testrun(suite: str, pr_number: int, sha: t.Optional[str] = None) -> bool:
```

File: scripts/needs_testrun.py [hunk 5]
```diff
@@ -181,15 +174,16 @@ def main() -> bool:
     argp = ArgumentParser()
 
     argp.add_argument("suite", help="The suite to use", type=str)
-    argp.add_argument("pr", help="The PR number", type=int)
+    argp.add_argument("--pr", help="The PR number", type=int, default=_get_pr_number())
+    argp.add_argument("--sha", help="Commit hash to use as diff base (defaults to PR merge root)", type=str)
```
Comment (P403n1x87) [inline scripts/needs_testrun.py:178]:
```suggestion
    argp.add_argument("--sha", help="Commit hash to use as diff base (defaults to PR merge root)", type=lambda v: v or None)
```

**Review by P403n1x87 — COMMENTED** [review]
Cool!

**Review by Yun-Kim — APPROVED** [review]
Awesome mechanism!

**Pull request merged**